pkg/render/applicationlayer.go

```diff
@@ -0,0 +1,308 @@
+	"github.com/tigera/operator/pkg/ptr"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/components"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/secret"
+	"github.com/tigera/operator/pkg/render/common/statik"
+)
+
+const (
+	CalicoSystemNamespace       = "calico-system"
+	EnvoyConfigConfigMapName    = "envoy-config"
+	L7LogCollectorDeamonsetName = "l7-log-collector"
+	L7CollectorContainerName    = "l7-collector"
+	ProxyContainerName          = "envoy-proxy"
+	EnvoyLogsKey                = "envoy-logs"
+	EnvoyConfigKey              = "envoy-config"
+	EnvoyConfigMapKey           = "envoy-config.yaml"
+)
+
+func ApplicationLayer(pullSecrets []*corev1.Secret, installation *operatorv1.InstallationSpec, osType rmeta.OSType,
+	applicationLayer *operatorv1.ApplicationLayer, l7Spec *L7LogCollectionSpec) Component {
+
+	return &ApplicationLayerComponent{
+		pullSecrets:      pullSecrets,
+		installation:     installation,
+		osType:           osType,
+		applicationLayer: applicationLayer,
+		l7Spec:           l7Spec,
+	}
+}
+
+type L7LogCollectionSpec struct {
+	Enabled                bool
+	LogIntervalSeconds     *int64
+	LogRequestsPerInterval *int64
+}
+
+type ApplicationLayerComponent struct {
+	pullSecrets      []*corev1.Secret
+	installation     *operatorv1.InstallationSpec
+	osType           rmeta.OSType
+	applicationLayer *operatorv1.ApplicationLayer
+	l7Spec           *L7LogCollectionSpec
+	proxyImage       string
+	collectorImage   string
+}
+
+func (c *ApplicationLayerComponent) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := c.installation.Registry
+	path := c.installation.ImagePath
+	prefix := c.installation.ImagePrefix
+
+	if c.osType != c.SupportedOSType() {
+		return fmt.Errorf("l7 log collection is supported only on %s", c.SupportedOSType())
+	}
+
+	var err error
+	var errMsgs []string
+
+	c.proxyImage, err = components.GetReference(components.ComponentEnvoyProxy, reg, path, prefix, is)
+
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+	c.collectorImage, err = components.GetReference(components.ComponentL7Collector, reg, path, prefix, is)
+
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+
+	if len(errMsgs) != 0 {
+		return fmt.Errorf(strings.Join(errMsgs, ","))
+	}
+
+	return nil
+}
+
+func (c *ApplicationLayerComponent) SupportedOSType() rmeta.OSType {
+	return rmeta.OSTypeLinux
+}
+
+func (c *ApplicationLayerComponent) Objects() ([]client.Object, []client.Object) {
+	var objs []client.Object
+	// TODO: ensure that namespace exists before proceeding
+	// if l7spec is provided render the required objects
+	if c.l7Spec != nil && c.l7Spec.Enabled {
+		objs = append(objs, c.envoyL7ConfigMap())
```

Comment:
I also wonder if this will need a PodSecurityPolicy on Openshift like calico-node.
Have you tried on Openshift?

Reply:
Also looking at node, I think you might need admin role binding for Docker EE.
https://github.com/tigera/operator/blob/14b2e7ca82c64b94ebb9e420efd67980fee62c30/pkg/render/node.go#L568